apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -509,98 +513,148 @@ private void ShutdownIfCompleted()
+        private void QueueCurrentBatchIfRequired(int extraByteCount = 0)
         {
             // Pre-condition: Must be locked
-            if (_currentBatch != null && (force || CurrentBatchIsFull(_idleClients.Count == 0, extraByteCount)))
+            if (_currentBatch == null)
             {
-                // Cancel the timeout for this batch.
-                _currentBatch.TimerCts.Cancel();
-                // Queue the batch ready for sending.
-                _batchesReady.Enqueue(_currentBatch);
-                // Mark that there is no current batch.
-                _currentBatch = null;
-                // Trigger send to server.
-                Task unusedSendTask = TriggerSend();
+                return;
+            }
+            // Current batch is full if either:
+            // * The number of messages in the batch >= the maximum number of messages allowed; or
+            // * The byte-count in the batch >= the maximum number of messages allowed.
+            // Special cases:
+            // * Before a message is queued, this method is called with the message byte-count;
+            //   If this message would cause the batch to exceed the maxmium byte-count, then this
+            //   batch is considered already full.
+            // * But if that is the first message in the batch, then that one message only is allowed
+            //   to make the batch go over its maximum allowed byte-count.
+            // The maximum size depends on whether local queueing is occuring.
+            // If _idleClients is empty, then all clients are currently sending, so local queueing is occuring;
+            // which means the current batch cannot be sent, even if it is full.
+            bool currentBatchIsFull = _idleClients.Count == 0 ?
+            _currentBatch.Messages.Count >= _batchMaxElementCount ||
+                (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount + extraByteCount >= _batchMaxByteCount) :
+            _currentBatch.Messages.Count >= _batchElementCountThreshold ||
+                (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount + extraByteCount >= _batchRequestByteThreshold);
+            if (currentBatchIsFull)
+            {
+                QueueCurrentBatch();
             }
         }
 
-        // Pre-condition: Must be locked
-        // Allow one message that is larger than requestByteThreshold
-        private bool CurrentBatchIsFull(bool isQueuing, int extraByteCount) => isQueuing ?
-                _currentBatch.Messages.Count >= _batchMaxElementCount || (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount >= _batchMaxByteCount) :
-                _currentBatch.Messages.Count >= _batchElementCountThreshold || (_currentBatch.Messages.Count > 0 && _currentBatch.ByteCount >= _batchRequestByteThreshold);
+        private void QueueCurrentBatch()
+        {
+            // Pre-condition: Must be locked
```

Comment:
I wonder whether we should add validation for these preconditions. I'm not sure how though... there's no Monitor.IsLocked :(